### PR TITLE
FormTokenField: Deprecate bottom margin

### PR DESCRIPTION
**Pull request by mirka**
Part of #39358

## What?

Officially start the deprecation period for the bottom margin on `FormTokenField`. And by extension, also on `PostTaxonomiesFlatTermSelector` in `@wordpress/editor`.

## Why?

All the unwanted margins on non-BaseControl-based components have already been hard deprecated, but this one was straggling.

## How?

The dependants of FormTokenField in this repo are the following:

- FormTokenField
  - Query Loop block
  - `wp.editor.PostTaxonomiesFlatTermSelector` (publicly exported as component)
    - `wp.editor.PostTaxonomies` (publicly exported as component)

Out of these, changing FormTokenField margins would only affect third-party consumers in the case of  `PostTaxonomiesFlatTermSelector`. So for that component, we are also doing a deprecation.

### Query Loop block

You may need this diff to show all the "Filters" options in the block inspector sidebar:

```diff
diff --git a/packages/block-library/src/query/edit/inspector-controls/index.js b/packages/block-library/src/query/edit/inspector-controls/index.js
index e71b2d09af..0c4471f187 100644
--- a/packages/block-library/src/query/edit/inspector-controls/index.js
+++ b/packages/block-library/src/query/edit/inspector-controls/index.js
@@ -268,7 +268,7 @@ export default function QueryInspectorControls( props ) {
 							/>
 						</ToolsPanelItem>
 					) }
-					{ showParentControl && (
+					{ true && (
 						<ToolsPanelItem
 							hasValue={ () => !! parents?.length }
 							label={ __( 'Parents' ) }
```

| Before | After |
|--------|-------|
|Query Loop block inspector controls, before|Query Loop block inspector controls, after|

The gutters in between used to be uneven and wrong. It is now uniform and [correct](https://github.com/WordPress/gutenberg/issues/43423#issuecomment-1232540038) (16px).

### PostTaxonomiesFlatTermSelector

In Post sidebar:

| Before | After |
|--------|-------|
|FlatTermSelector with extra margin|FlatTermSelector with no margin|

There used to be an extra bottom margin. Now it is gone.

---

## ✍️ Dev Note

See Dev Note in #39358.

### CODE REVIEW

File: packages/interface/src/components/complementary-area/style.scss [hunk 1]
```diff
@@ -33,7 +33,7 @@
 		}
 	}
 
-	p:not(.components-base-control__help) {
+	p:not(.components-base-control__help, .components-form-token-field__help) {
 		margin-top: 0;
 	}
```
Comment (mirka) [inline packages/interface/src/components/complementary-area/style.scss:38]: This is to maintain this margin right here, which was previously maintained implicitly due to a `Spacer` component inserted in the `__nextHasNoMarginBottom={ false }` case.

Margin between FormTokenField input field and help text, in sidebar

File: packages/editor/src/components/post-taxonomies/index.js
Comment (mirka) [inline packages/editor/src/components/post-taxonomies/index.js:1]: The `PostTaxonomies` component is used in two places:

- The "Tags" section of the Post sidebar.

    Tags section

- In the Preferences modal.
 
    Toggles in the Preferences modal

https://github.com/WordPress/gutenberg/blob/9f8014ad8aea4ed4602dca039c1c89624a01c330/packages/editor/src/components/preferences-modal/index.js#L105-L111

The usage in the Preferences modal is very strange — the callback parameters are used to render toggles 😅 Anyway, our margin changes don't affect this use case.

File: packages/editor/src/components/post-publish-panel/maybe-tags-panel.js
Comment (mirka) [inline packages/editor/src/components/post-publish-panel/maybe-tags-panel.js:1]: This is in the Post publish panel:

Tags input in Post publish panel

**Comment by tyxla** [comment]
> You may need this diff to show all the "Filters" options in the block inspector sidebar:

No need to do that. Just make sure to test with both a hierarchical post type (page) and a flat one (post). You can select it from the Query Loop block settings:

Screenshot 2024-07-15 at 14 40 21

File: packages/editor/src/components/post-taxonomies/flat-term-selector.js [hunk 2]
```diff
@@ -55,16 +56,28 @@ const termNamesToIds = ( names, terms ) => {
...
+export function FlatTermSelector( { slug, __nextHasNoMarginBottom } ) {
 	const [ values, setValues ] = useState( [] );
 	const [ search, setSearch ] = useState( '' );
 	const debouncedSearch = useDebounce( setSearch, 500 );
 
+	if ( ! __nextHasNoMarginBottom ) {
+		deprecated(
+			'Bottom margin styles for wp.editor.PostTaxonomiesFlatTermSelector',
+			{
+				since: '6.7',
```
Comment (tyxla) [inline packages/editor/src/components/post-taxonomies/flat-term-selector.js:74]: Is this going to land in 6.7 actually?

File: packages/block-library/src/query/editor.scss [hunk 3]
```diff
@@ -49,15 +49,6 @@
 	}
 }
 
-.block-library-query-toolspanel__filters {
```
Comment (tyxla) [inline packages/block-library/src/query/editor.scss:52]: We're still rendering the original classname for backward compatibility, right?

File: packages/components/src/form-token-field/stories/index.story.tsx [hunk 4]
```diff
@@ -99,6 +100,7 @@ export const Async: StoryFn< typeof FormTokenField > = ( {
 Async.args = {
 	label: 'Type a continent',
 	suggestions: continents,
+	__nextHasNoMarginBottom: true,
```
Comment (tyxla) [inline packages/components/src/form-token-field/stories/index.story.tsx:103]: Confirmed that all stories inherit args from either of the 2 definitions that define `__nextHasNoMarginBottom: true`.

**Review by tyxla — APPROVED** [review]
I spent some time testing thoroughly in Storybook and in the post/site editor, everything tests well. ✅ 

I also took some time to confirm all instances in the codebase have been addressed. 👍 

LGTM 🚀

File: packages/editor/src/components/post-taxonomies/flat-term-selector.js (hunk 2 above)
Comment (mirka) [reply packages/editor/src/components/post-taxonomies/flat-term-selector.js:74]: Yes, too late to land in 6.6.

File: packages/block-library/src/query/editor.scss (hunk 3 above)
Comment (mirka) [reply packages/block-library/src/query/editor.scss:52]: Correct. I added a code comment b912e8b

**Pull request merged**